jOOQ/src/main/java/org/jooq/impl/DefaultRecordMapper.java

```diff
@@ -343,6 +349,33 @@ private final void init(E instance) {
             return;
         }
 
+        // [#7324] Map Kotlin data classes
+        KClass<? extends E> kotlinClass = JvmClassMappingKt.getKotlinClass(type);
+        if (kotlinClass.isData()) {
+            KFunction<? extends E> primaryConstructor = KClasses.getPrimaryConstructor(kotlinClass);
+            List<KParameter> parameters = primaryConstructor.getParameters();
+
+            if (parameters.size() != fields.length) {
+                throw new MappingException("Primary constructor of Kotlin data class " + type + " did not match row type " + rowType);
+            }
```

Comment:
Once we reflect on field names, I don't think this limitation should still be maintained. A record may have more or less fields than the type we're mapping it into. See also the Java 8 parameter name reflection implementation, as an example.